File: routes/admin/ads.py
"""
UrgenceMaroc.com
By MOA Digital Agency LLC
Developed by: Aisance KALONJI
Contact: moa@myoneart.com
Website: www.myoneart.com

routes/admin/ads.py - Gestion des publicités
Ce fichier gère le CRUD des publicités (images/vidéos) et les paramètres
de déclenchement du système publicitaire.
"""

from flask import render_template, request, redirect, url_for, flash, jsonify
from flask_login import login_required
from models.advertisement import Advertisement, AdSettings
from extensions import db, csrf
from datetime import datetime
from routes.admin import admin_bp, allowed_file, get_upload_path, safe_delete_upload, save_upload_file


@admin_bp.route('/ads')
@login_required
def list_ads():
    ads = Advertisement.query.order_by(Advertisement.priority.desc(), Advertisement.created_at.desc()).all()
    return render_template('admin/ads.html', ads=ads)


@admin_bp.route('/ad/add', methods=['GET', 'POST'])
@login_required
def add_ad():
    if request.method == 'POST':
        image_filename = None
        if request.form.get('media_type') == 'image' and 'image_file' in request.files:
            file = request.files['image_file']
            image_filename = save_upload_file(file, 'ads', '')
        
        start_date = None
        end_date = None
        if request.form.get('start_date'):
            start_date = datetime.strptime(request.form.get('start_date'), '%Y-%m-%dT%H:%M')
        if request.form.get('end_date'):
            end_date = datetime.strptime(request.form.get('end_date'), '%Y-%m-%dT%H:%M')
        
        ad = Advertisement(
            title=request.form.get('title'),
            description=request.form.get('description', ''),
            media_type=request.form.get('media_type', 'image'),
            image_filename=image_filename,
            video_url=request.form.get('video_url', ''),
            cta_text=request.form.get('cta_text', 'En savoir plus'),
            cta_url=request.form.get('cta_url', ''),
            skip_delay=int(request.form.get('skip_delay', 5)) or 5,
            priority=int(request.form.get('priority', 0)),
            start_date=start_date,
            end_date=end_date,
            is_active=request.form.get('is_active') == 'on'
        )
        db.session.add(ad)
        db.session.commit()
        flash('Publicité ajoutée avec succès', 'success')
        return redirect(url_for('admin.list_ads'))
    
    return render_template('admin/ad_form.html', ad=None)


@admin_bp.route('/ad/<int:id>/edit', methods=['GET', 'POST'])
@login_required
def edit_ad(id):
    ad = Advertisement.query.get_or_404(id)
    
    if request.method == 'POST':
        if request.form.get('media_type') == 'image':
            if 'image_file' in request.files:
                file = request.files['image_file']
                if file and file.filename and allowed_file(file.filename):
                    if ad.image_filename:
                        safe_delete_upload(ad.image_filename, 'ads')
                    ad.image_filename = save_upload_file(file, 'ads', '')
            
            if request.form.get('remove_image') == 'on':
                if ad.image_filename:
                    safe_delete_upload(ad.image_filename, 'ads')
                    ad.image_filename = None
        
        start_date = None
        end_date = None
        if request.form.get('start_date'):
            start_date = datetime.strptime(request.form.get('start_date'), '%Y-%m-%dT%H:%M')
        if request.form.get('end_date'):
            end_date = datetime.strptime(request.form.get('end_date'), '%Y-%m-%dT%H:%M')
        
        ad.title = request.form.get('title')
        ad.description = request.form.get('description', '')
        ad.media_type = request.form.get('media_type', 'image')
        ad.video_url = request.form.get('video_url', '')
        ad.cta_text = request.form.get('cta_text', 'En savoir plus')
        ad.cta_url = request.form.get('cta_url', '')
        ad.skip_delay = int(request.form.get('skip_delay', 5)) or 5
        ad.priority = int(request.form.get('priority', 0))
        ad.start_date = start_date
        ad.end_date = end_date
        ad.is_active = request.form.get('is_active') == 'on'
        
        db.session.commit()
        flash('Publicité mise à jour', 'success')
        return redirect(url_for('admin.list_ads'))
    
    return render_template('admin/ad_form.html', ad=ad)


@admin_bp.route('/ad/<int:id>/toggle', methods=['POST'])
@login_required
@csrf.exempt
def toggle_ad(id):
    ad = Advertisement.query.get_or_404(id)
    ad.is_active = not ad.is_active
    db.session.commit()
    return jsonify({'success': True, 'is_active': ad.is_active})


@admin_bp.route('/ad/<int:id>/delete', methods=['POST'])
@login_required
def delete_ad(id):
    ad = Advertisement.query.get_or_404(id)
    if ad.image_filename:
        safe_delete_upload(ad.image_filename, 'ads')
    db.session.delete(ad)
    db.session.commit()
    flash('Publicité supprimée', 'success')
    return redirect(url_for('admin.list_ads'))


@admin_bp.route('/ads/settings', methods=['GET', 'POST'])
@login_required
def ad_settings():
    settings = AdSettings.get_settings()
    
    if request.method == 'POST':
        settings.ads_enabled = request.form.get('ads_enabled') == 'on'
        settings.trigger_type = request.form.get('trigger_type', 'time')
        settings.time_delay = int(request.form.get('time_delay', 60))
        settings.time_repeat = request.form.get('time_repeat') == 'on'
        settings.time_interval = int(request.form.get('time_interval', 300))
        settings.page_count = int(request.form.get('page_count', 3))
        settings.refresh_show = request.form.get('refresh_show') == 'on'
        settings.refresh_count = int(request.form.get('refresh_count', 1))
        settings.default_skip_delay = int(request.form.get('default_skip_delay', 5))
        settings.max_ads_per_session = int(request.form.get('max_ads_per_session', 10))
        settings.cooldown_after_skip = int(request.form.get('cooldown_after_skip', 60))
        settings.cooldown_after_click = int(request.form.get('cooldown_after_click', 300))
        settings.show_on_mobile = request.form.get('show_on_mobile') == 'on'
        settings.show_on_desktop = request.form.get('show_on_desktop') == 'on'
        
        db.session.commit()
        flash('Paramètres enregistrés avec succès', 'success')
        return redirect(url_for('admin.ad_settings'))
    
    return render_template('admin/ad_settings.html', settings=settings)

File: static/js/ads.js
/**
 * UrgenceMaroc.com
 * By MOA Digital Agency LLC
 * Developed by: Aisance KALONJI
 * Contact: moa@myoneart.com
 * Website: www.myoneart.com
 *
 * static/js/ads.js - Système publicitaire
 * Ce fichier gère le système de publicités: déclenchement par temps/pages,
 * affichage des pubs (image/vidéo), compteurs et cooldowns.
 */

let adSettings = null;
let adState = {
    adsShown: parseInt(sessionStorage.getItem('adAdsShown') || '0'),
    pageChanges: parseInt(sessionStorage.getItem('adPageChanges') || '0'),
    refreshCount: parseInt(sessionStorage.getItem('adRefreshCount') || '0'),
    lastAdTime: parseInt(sessionStorage.getItem('adLastAdTime') || '0'),
    cooldownUntil: parseInt(sessionStorage.getItem('adCooldownUntil') || '0'),
    currentAdId: null,
    countdownTimer: null,
    timeTimer: null
};

function saveAdState() {
    sessionStorage.setItem('adAdsShown', adState.adsShown.toString());
    sessionStorage.setItem('adPageChanges', adState.pageChanges.toString());
    sessionStorage.setItem('adRefreshCount', adState.refreshCount.toString());
    sessionStorage.setItem('adLastAdTime', adState.lastAdTime.toString());
    sessionStorage.setItem('adCooldownUntil', adState.cooldownUntil.toString());
}

async function initAdSystem() {
    try {
        const response = await fetch('/api/ads/settings');
        adSettings = await response.json();
        
        if (!adSettings.ads_enabled) return;
        
        const isMobile = window.innerWidth < 768;
        if (isMobile && !adSettings.show_on_mobile) return;
        if (!isMobile && !adSettings.show_on_desktop) return;
        
        if (adSettings.trigger_type === 'refresh' || adSettings.trigger_type === 'combined') {
            adState.refreshCount++;
            sessionStorage.setItem('adRefreshCount', adState.refreshCount.toString());
            
            if (adSettings.refresh_show && adState.refreshCount >= adSettings.refresh_count) {
                adState.refreshCount = 0;
                saveAdState();
                setTimeout(() => showRandomAd(), 1000);
                return;
            }
        }
        
        saveAdState();
        
        if (adSettings.trigger_type === 'time' || adSettings.trigger_type === 'combined') {
            adState.timeTimer = setTimeout(() => {
                showRandomAd();
                if (adSettings.time_repeat) {
                    startAdInterval();
                }
            }, adSettings.time_delay * 1000);
        }
        
    } catch (error) {
        console.log('Ad system not available');
    }
}

function startAdInterval() {
    if (adState.timeTimer) clearInterval(adState.timeTimer);
    adState.timeTimer = setInterval(() => {
        if (canShowAd()) {
            showRandomAd();
        }
    }, adSettings.time_interval * 1000);
}

function canShowAd() {
    if (!adSettings || !adSettings.ads_enabled) return false;
    if (adState.adsShown >= adSettings.max_ads_per_session) return false;
    if (Date.now() < adState.cooldownUntil) return false;
    
    const adModal = document.getElementById('adModal');
    if (adModal && adModal.classList.contains('flex')) return false;
    
    return true;
}

function trackPageChange() {
    if (!adSettings) return;
    if (adSettings.trigger_type !== 'page' && adSettings.trigger_type !== 'combined') return;
    
    adState.pageChanges++;
    
    if (adState.pageChanges >= adSettings.page_count && canShowAd()) {
        adState.pageChanges = 0;
        saveAdState();
        showRandomAd();
    } else {
        saveAdState();
    }
}

async function showRandomAd() {
    if (!canShowAd()) return;
    
    try {
        const response = await fetch('/api/ads/random');
        const ad = await response.json();
        
        if (!ad) return;
        
        adState.currentAdId = ad.id;
        displayAd(ad);
        
        fetch(`/api/ads/${ad.id}/view`, { method: 'POST' })
            .catch(err => console.log('Failed to record ad view'));
        
    } catch (error) {
        console.log('Failed to load ad');
    }
}

function displayAd(ad) {
    const modal = document.getElementById('adModal');
    const mediaContainer = document.getElementById('adMediaContainer');
    const title = document.getElementById('adTitle');
    const description = document.getElementById('adDescription');
    const ctaBtn = document.getElementById('adCtaBtn');
    const skipBtn = document.getElementById('adSkipBtn');
    const countdown = document.getElementById('adCountdown');
    
    title.textContent = ad.title;
    description.textContent = ad.description || '';
    ctaBtn.textContent = ad.cta_text || 'En savoir plus';
    ctaBtn.href = ad.cta_url || '#';
    
    if (!ad.cta_url) {
        ctaBtn.classList.add('hidden');
    } else {
        ctaBtn.classList.remove('hidden');
    }
    
    mediaContainer.innerHTML = '';
    if (ad.media_type === 'image' && ad.image_url) {
        const img = document.createElement('img');
        img.src = ad.image_url;
        img.alt = ad.title;
        img.className = 'w-full h-48 object-cover';
        mediaContainer.appendChild(img);
    } else if (ad.media_type === 'video' && ad.video_url) {
        const embedUrl = getVideoEmbedUrl(ad.video_url);
        if (embedUrl) {
            const iframe = document.createElement('iframe');
            iframe.src = embedUrl;
            iframe.className = 'w-full h-56';
            iframe.frameBorder = '0';
            iframe.allowFullscreen = true;
            iframe.allow = 'autoplay; encrypted-media';
            mediaContainer.appendChild(iframe);
        }
    }
    
    let timeLeft = ad.skip_delay || 5;
    countdown.textContent = timeLeft;
    skipBtn.disabled = true;
    skipBtn.classList.add('cursor-not-allowed', 'bg-gray-200', 'text-gray-500');
    skipBtn.classList.remove('cursor-pointer', 'bg-gray-300', 'text-gray-700', 'hover:bg-gray-400');
    
    if (adState.countdownTimer) clearInterval(adState.countdownTimer);
    
    adState.countdownTimer = setInterval(() => {
        timeLeft--;
        countdown.textContent = timeLeft;
        
        if (timeLeft <= 0) {
            clearInterval(adState.countdownTimer);
            skipBtn.disabled = false;
            skipBtn.textContent = 'Passer la pub';
            skipBtn.classList.remove('cursor-not-allowed', 'bg-gray-200', 'text-gray-500');
            skipBtn.classList.add('cursor-pointer', 'bg-gray-300', 'text-gray-700', 'hover:bg-gray-400');
        }
    }, 1000);
    
    modal.classList.remove('hidden');
    modal.classList.add('flex');
    adState.adsShown++;
    adState.lastAdTime = Date.now();
    saveAdState();
}

function getVideoEmbedUrl(url) {
    const ytMatch = url.match(/(?:youtube\.com\/watch\?v=|youtu\.be\/|youtube\.com\/embed\/)([^&\s]+)/);
    if (ytMatch) {
        return `https://www.youtube.com/embed/${ytMatch[1]}?autoplay=1`;
    }
    
    if (url.includes('facebook.com')) {
        return `https://www.facebook.com/plugins/video.php?href=${encodeURIComponent(url)}&show_text=false&autoplay=true`;
    }
    
    const vimeoMatch = url.match(/vimeo\.com\/(\d+)/);
    if (vimeoMatch) {
        return `https://player.vimeo.com/video/${vimeoMatch[1]}?autoplay=1`;
    }
    
    return null;
}

function closeAdModal() {
    const modal = document.getElementById('adModal');
    modal.classList.add('hidden');
    modal.classList.remove('flex');
    
    if (adState.countdownTimer) {
        clearInterval(adState.countdownTimer);
    }
    
    if (adSettings) {
        adState.cooldownUntil = Date.now() + (adSettings.cooldown_after_skip * 1000);
        saveAdState();
    }
    
    adState.currentAdId = null;
}

function handleAdClick() {
    if (adState.currentAdId) {
        fetch(`/api/ads/${adState.currentAdId}/click`, { method: 'POST' })
            .catch(err => console.log('Failed to record ad click'));
    }
    
    if (adSettings) {
        adState.cooldownUntil = Date.now() + (adSettings.cooldown_after_click * 1000);
        saveAdState();
    }
    
    setTimeout(closeAdModal, 100);
}

// Expose to global scope
window.initAdSystem = initAdSystem;
window.showRandomAd = showRandomAd;
window.displayAd = displayAd;
window.closeAdModal = closeAdModal;
window.handleAdClick = handleAdClick;
window.trackPageChange = trackPageChange;
